core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -659,6 +669,47 @@ class LogManager(logDirs: Seq[File],
       Option(currentLogs.get(topicPartition))
   }
 
+  /**
+   * Method to indicate that logs are getting initialized for the partition passed in as argument.
+   * This method should always be followed by [[kafka.log.LogManager#finishedInitializingLog]] to indicate that log
+   * initialization is done.
+   */
+  def initializingLog(topicPartition: TopicPartition): Unit = configUpdateLock synchronized {
+    partitionsInitializing(topicPartition) = false
+  }
+
+  /**
+   * Mark the partition configuration for partitions that are getting initialized for topic
+   * as dirty. That will result in reloading of configuration once initialization is done.
+   */
+  def topicConfigUpdated(topic: String): Unit = configUpdateLock synchronized {
+    partitionsInitializing.collect {
+      case (topicPartition, _) if topicPartition.topic() == topic => partitionsInitializing(topicPartition) = true
+    }
+  }
+
+  /**
+   * Mark all in progress partitions having dirty configuration if broker configuration is updated.
+   */
+  def brokerConfigUpdated(): Unit = configUpdateLock synchronized {
+    partitionsInitializing.foreach {
+      case (topicPartition, _) => partitionsInitializing(topicPartition) = true
+    }
+  }
+
+  /**
+   * Method to indicate that the log initialization for the partition passed in as argument is
+   * finished. This method should follow a call to [[kafka.log.LogManager#initializingLog]]
+   */
+  def finishedInitializingLog(topicPartition: TopicPartition,
+                              log: Option[Log],
+                              config: => LogConfig): Unit = configUpdateLock synchronized {
+    if (partitionsInitializing(topicPartition)) {
+      log.foreach(_.config = config)
```

Comment:
Seems undesirable to be holding the lock while fetching the topic config. Would it be possible to let this return a boolean indicating whether the config needs to be reinitialized rather than doing it in place?

Reply:
Updated to use ConcurrentHashMap. Now the code checks and refreshes config in a loop to take care of race.